_posts/2020-09-15-mutiny-pagination.adoc

```diff
@@ -0,0 +1,200 @@
+That's fun, and even better, it uses pagination. 
+We got our API!
+
+If you call `https://api.punkapi.com/v2/beers?page=1`, you get a JSON array like:
+
+[source, json]
+----
+[
+    {
+        first beer
+    },
+    {
+        second beer
+    }, 
+    // ...
+]
+----
+
+I won't details the content of each object, the documentation page does a great job about that.
+Let's focus on the pagination aspect.
+First, we passed the `page` query parameter, indicating which page we want.
+Generally, when you retrieve a page, the API provides a way to know if there is a next page (a special field in the JSON document, or HTTP header), but the Punk API does not provide any hint. 
+So, to retrieve all the beers, we need to invoke the service for page 1, 2, 3... until the returned JSON array is empty. 
+
+In an imperative word, to retrieve all the beers, you would do something like this:
+
+[source, java]
+----
+List<Beer> beers = ...;
+int page = 1;
+List<Beer> batch = ...
+do {
+  batch= getBeersFromPage(page);
+  beers.addAll(batch);
+  page = page + 1;
+} while (! batch.isEmpty());
+----
+
+How can we achieve the same in a reactive manner and build a stream of beer?
+
+image:/assets/images/posts/mutiny-pagination/mutiny-pagination.png[]
+
+Let's proceed step by step.
+
+## Retrieving a single page
+
+First, we need to see how we can retrieve a single page.
+I'm going to use the Vert.x Web Client, but you can use any reactive HTTP clients providing a Mutiny API.
+
+[source, java]
+----
+// Create the client
+WebClient client = WebClient.create(vertx, new WebClientOptions()
+      .setDefaultHost("api.punkapi.com")
+      .setDefaultPort(443)
+      .setSsl(true)
+);
+
+// Retrieve the first page
+Uni<List<Beer>> uni = client.get("/v2/beers?page=1")
+      .send()
+      .onItem().transform(Pagination::toListOfBeer);
+----
+
+This snippet creates the web client. 
+Then, we use that client and retrieve the first page.
+
+When we receive the result (`onItem`), we transform the JSON array into a list of beers.
+
+Let's extract this code in a method and take the page number as parameter:
+
+[source, java]
+----
+private static Uni<List<Beer>> getPage(WebClient client, int page) {
+    return client.get("/v2/beers?page=" + page)
+            .send()
+            .onItem().transform(Pagination::toListOfBeer);
+}
+----
+
+So far, so good.
+
+## Retrieving multiple page
+
+So, now, we know how to retrieve a single page and extract the items from it. 
+We just need to repeat this operation for every page, and provide a stream.
+
+Mutiny provides a method to create a `Multi` by repeating multiple times a `Uni`.
+Under the hood, it calls a method returning a `Uni` and subscribe on it.
+But we need to make _progress_, and pass the current page.
+Mutiny offers the possibility to store a state, to the method creating the `Uni` can increment and get the page number:
```

Comment:
I think ", to the method" should be ", so the method"?

Reply:
Replacing it with: `Mutiny offers the possibility to store a state in order to let the method creating the `Uni` increments the page number:`